src/cycles/simple_cycles.c

```diff
@@ -0,0 +1,314 @@
+                        if (!discrepancy_found)
+                        {
+                            // found this loop already.
+                            duplicate_found = true;
+                            break;
+                        }
+                    }
+                    if (duplicate_found)
+                    {
+                        persist_result = false;
+                    }
+                }
+                // end filter
+                if (persist_result) {
+                    IGRAPH_CHECK(igraph_vector_int_list_push_back_copy(results, &res));
+                }
+                igraph_vector_int_destroy(&res);
+                IGRAPH_FINALLY_CLEAN(1);
+            }
+        }
+        else if (!(VECTOR(state->blocked)[W]))
+        {
+            IGRAPH_CHECK(igraph_simple_cycles_circuit(state, W, S, results, &local_found, search_mode));
+        }
+    }
+    *found = local_found;
+
+    // L2
+    if (local_found)
+    {
+        IGRAPH_CHECK(igraph_simple_cycles_unblock(state, V));
+    }
+    else
+    {
+        for (igraph_integer_t i = 0; i < igraph_vector_int_size(igraph_adjlist_get(&state->AK, V)); ++i)
+        {
+            igraph_integer_t W = VECTOR(*igraph_adjlist_get(&state->AK, V))[i];
+            igraph_integer_t pos;
+            if (!igraph_vector_int_search(igraph_adjlist_get(&state->B, W), 0, V, &pos))
+            {
+                IGRAPH_CHECK(igraph_vector_int_push_back(igraph_adjlist_get(&state->B, W), V));
+            }
+        }
+    }
+
+    if (igraph_stack_int_size(&state->stack) > 0)
+    {
+        // unstack v
+        // printf("Unstacking %lld\n", V);
+        igraph_stack_int_pop(&state->stack);
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_simple_cycle_search_state_init
+ * \brief Initialize the cycle search state
+ *
+ * \param state The state structure to initialize
+ * \param graph The graph object
+ * \return Error code.
+ *
+ * Time complexity: O(|V|*|E|*log(|V|*|E|))
+ *
+ * \ref igraph_simple_cycle_search_state_destroy
+ */
+igraph_error_t igraph_simple_cycle_search_state_init(igraph_simple_cycle_search_state_t *state, const igraph_t *graph)
+{
+    igraph_integer_t N = igraph_vcount(graph);
+
+    state->N = N;
+    IGRAPH_CHECK(igraph_stack_int_init(&state->stack, N)); // maximum size per cycle.
+    IGRAPH_FINALLY(igraph_stack_int_destroy, &state->stack);
+    IGRAPH_CHECK(igraph_vector_bool_init(&state->blocked, N));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &state->blocked);
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &state->AK, IGRAPH_OUT, IGRAPH_LOOPS_ONCE, IGRAPH_MULTIPLE)); // TODO: understand what we actually want to include
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &state->AK);
+    igraph_adjlist_sort(&state->AK);
+    state->directed = igraph_is_directed(graph);
+    IGRAPH_CHECK(igraph_adjlist_init_empty(&state->B, N));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &state->B);
+
+    return IGRAPH_SUCCESS;
+}
```

Comment:
There appears to be a problem here. When a constructor like this returns, it must clean up the finally stack. Then, at the point where the constructor is called, the `igraph_simple_cycle_search_state_t` structure must be placed on the finally stack with its destructor.